File: bridge/clients/myfb-bridge.mjs
/**
 * My-Feedbacks DB Bridge — Node.js client (ESM, zero deps).
 *
 * Mirror of the JS extension client (sidepanel/db-bridge-client.js), the
 * Python client (myfb_bridge.py) and the Go client (myfb_bridge.go).
 * Same HMAC math, same canonical-args quirk (PHP compat — empty args
 * serialize to "[]"), same humanized errors.
 *
 * Node 18+ (native fetch + node:crypto.webcrypto). Also works in Deno
 * and Bun as-is. NO dependencies.
 *
 * Usage:
 *
 *   import { BridgeClient } from './myfb-bridge.mjs';
 *
 *   const bridge = new BridgeClient({
 *     url:    'https://example.com/myfb-bridge.php',
 *     secret: process.env.BRIDGE_SECRET,
 *   });
 *
 *   const { tableCount } = await bridge.meta();
 *   const tables         = await bridge.tables();
 *   const md             = await bridge.schemaMd();
 *
 * Errors throw `Error` with humanized French message. Catch and log.
 */

import { createHmac, randomBytes } from 'node:crypto';

export const MIN_BRIDGE_VERSION = '1.1.0';

// ── Signing ────────────────────────────────────────────────────────────

/** Match PHP's json_encode($args, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE). */
function canonArgs(args) {
  if (!args || (typeof args === 'object' && Object.keys(args).length === 0)) {
    return '[]';   // PHP encodes empty assoc as []
  }
  return JSON.stringify(args);
}

function signRequest(secret, ts, nonce, op, args) {
  const msg = `${ts}.${nonce}.${op}.${canonArgs(args)}`;
  return createHmac('sha256', secret).update(msg).digest('hex');
}

function randomNonce() {
  return randomBytes(16).toString('hex');
}

// ── Errors ─────────────────────────────────────────────────────────────

function humanize(status, raw) {
  raw = String(raw || '').toLowerCase();
  if (status === 0)   return 'Bridge inaccessible (réseau / DNS / CORS)';
  if (status === 404) return "Endpoint introuvable — vérifie l'URL du bridge";
  if (status === 401) {
    if (raw.includes('replay')) return 'Nonce déjà utilisé — horloge décalée ?';
    if (raw.includes('stale'))  return 'Horloge client/serveur décalée (> 60s)';
    if (raw.includes('nonce'))  return 'Nonce invalide';
    return 'Signature HMAC invalide — secret incorrect ?';
  }
  if (status === 403) return 'Table non exposée par la config du bridge';
  if (status === 405) return 'Méthode HTTP refusée — bridge mal configuré';
  if (status >= 500)  return 'Erreur interne du bridge — vérifie audit.log';
  return raw || `Erreur HTTP ${status}`;
}

function cmpVer(a, b) {
  const pa = String(a || '0').split('.').map((x) => parseInt(x, 10) || 0);
  const pb = String(b || '0').split('.').map((x) => parseInt(x, 10) || 0);
  for (let i = 0; i < Math.max(pa.length, pb.length); i++) {
    const va = pa[i] || 0, vb = pb[i] || 0;
    if (va < vb) return -1;
    if (va > vb) return 1;
  }
  return 0;
}

// ── Client ─────────────────────────────────────────────────────────────

export class BridgeClient {
  constructor({ url, secret, timeoutMs = 10_000 }) {
    if (!url || !secret) throw new Error('Bridge non configuré (URL et secret requis)');
    this.url       = url;
    this.secret    = secret;
    this.timeoutMs = timeoutMs;
  }

  async call(op, args = {}) {
    const ts    = Math.floor(Date.now() / 1000);
    const nonce = randomNonce();
    const body  = { op, args, ts, nonce, sig: signRequest(this.secret, ts, nonce, op, args) };

    const ctrl  = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), this.timeoutMs);
    let resp;
    try {
      resp = await fetch(this.url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
        signal: ctrl.signal,
      });
    } catch (e) {
      clearTimeout(timer);
      throw new Error(humanize(0, e?.message || String(e)));
    }
    clearTimeout(timer);

    const text = await resp.text();
    let parsed = null;
    try { parsed = JSON.parse(text); } catch { /* malformed */ }
    if (!parsed) throw new Error(humanize(resp.status, 'malformed response'));

    if (parsed.version && cmpVer(parsed.version, MIN_BRIDGE_VERSION) < 0) {
      throw new Error(
        `Bridge trop ancien (${parsed.version} < ${MIN_BRIDGE_VERSION}) — ` +
        `mets à jour myfb-bridge.php`,
      );
    }
    if (!parsed.ok) throw new Error(humanize(resp.status, parsed.error));
    return parsed.data ?? {};
  }

  meta()                     { return this.call('meta'); }
  tables()                   { return this.call('tables').then((d) => d.tables || []); }
  describe(table)            { return this.call('describe', { table }).then((d) => d.columns || []); }
  sample(table, opts = {})   {
    return this.call('sample', { table, limit: opts.limit ?? 3, strategy: opts.strategy || 'mixed' });
  }
  count(table)               { return this.call('count', { table }).then((d) => d.count | 0); }
  schemaMd()                 { return this.call('schema_md').then((d) => d.markdown || ''); }
}

// Helpers exported for tests / custom usage. Mirror the JS sidepanel API.
export const _internal = { signRequest, canonArgs, humanize, cmpVer, randomNonce };

File: bridge/clients/myfb_bridge.rb
# frozen_string_literal: true
#
# My-Feedbacks DB Bridge — Ruby client (stdlib only).
#
# Mirror of the JS extension client (sidepanel/db-bridge-client.js), the
# Python client (myfb_bridge.py), the Go client (myfb_bridge.go) and the
# Node client (myfb-bridge.mjs). Same HMAC math, same canonical-args
# quirk for PHP compat (empty args serialize to "[]", not "{}"), same
# humanized errors.
#
# Ruby 3.0+. Stdlib only — no Gemfile, no net-http-persistent, no faraday.
#
# Usage :
#
#   require_relative 'myfb_bridge'
#
#   bridge = MyFb::BridgeClient.new(
#     url:    'https://example.com/myfb-bridge.php',
#     secret: ENV.fetch('BRIDGE_SECRET'),
#   )
#
#   meta   = bridge.meta             # → Hash
#   tables = bridge.tables           # → Array<Hash>
#   md     = bridge.schema_md        # → String

require 'json'
require 'net/http'
require 'openssl'
require 'securerandom'
require 'uri'

module MyFb
  MIN_BRIDGE_VERSION = '1.1.0'

  class BridgeError < StandardError; end

  class BridgeClient
    def initialize(url:, secret:, timeout: 10)
      raise BridgeError, 'bridge non configuré (URL et secret requis)' if url.to_s.empty? || secret.to_s.empty?

      @uri     = URI.parse(url)
      @secret  = secret
      @timeout = timeout
    end

    # ── Operations ───────────────────────────────────────────────

    def meta;                            call('meta');                                end
    def tables;                          call('tables').fetch('tables', []);          end
    def describe(table);                 call('describe', { table: table }).fetch('columns', []); end
    def sample(table, limit: 3, strategy: 'mixed')
      call('sample', { table: table, limit: limit, strategy: strategy })
    end
    def count(table);                    call('count', { table: table }).fetch('count', 0).to_i; end
    def schema_md;                       call('schema_md').fetch('markdown', '');     end

    # ── Core ─────────────────────────────────────────────────────

    def call(op, args = {})
      ts    = Time.now.to_i
      nonce = SecureRandom.hex(16)
      body  = { op: op, args: args, ts: ts, nonce: nonce }
      body[:sig] = sign(ts, nonce, op, args)

      req = Net::HTTP::Post.new(@uri)
      req['Content-Type'] = 'application/json'
      req.body            = JSON.generate(body)

      http               = Net::HTTP.new(@uri.host, @uri.port)
      http.use_ssl       = @uri.scheme == 'https'
      http.open_timeout  = @timeout
      http.read_timeout  = @timeout

      begin
        resp = http.request(req)
      rescue StandardError => e
        raise BridgeError, humanize(0, e.message)
      end

      parsed = begin
        JSON.parse(resp.body)
      rescue JSON::ParserError
        raise BridgeError, humanize(resp.code.to_i, 'malformed response')
      end

      version = parsed['version']
      if version && cmp_ver(version, MIN_BRIDGE_VERSION) < 0
        raise BridgeError,
              "Bridge trop ancien (#{version} < #{MIN_BRIDGE_VERSION}) — " \
              'mets à jour myfb-bridge.php'
      end

      raise BridgeError, humanize(resp.code.to_i, parsed['error']) unless parsed['ok']

      parsed['data'] || {}
    end

    private

    # PHP's json_encode($args, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE).
    # Empty hash → "[]" (PHP encodes empty assoc as a JSON array).
    def canon_args(args)
      return '[]' if args.nil? || (args.is_a?(Hash) && args.empty?)

      JSON.generate(args)
    end

    def sign(ts, nonce, op, args)
      msg = "#{ts}.#{nonce}.#{op}.#{canon_args(args)}"
      OpenSSL::HMAC.hexdigest('SHA256', @secret, msg)
    end

    def humanize(status, raw)
      r = (raw || '').to_s.downcase
      return 'Bridge inaccessible (réseau / DNS / CORS)'           if status.zero?
      return "Endpoint introuvable — vérifie l'URL du bridge"      if status == 404
      if status == 401
        return 'Nonce déjà utilisé — horloge décalée ?'            if r.include?('replay')
        return 'Horloge client/serveur décalée (> 60s)'            if r.include?('stale')
        return 'Nonce invalide'                                    if r.include?('nonce')

        return 'Signature HMAC invalide — secret incorrect ?'
      end
      return 'Table non exposée par la config du bridge'           if status == 403
      return 'Méthode HTTP refusée — bridge mal configuré'         if status == 405
      return 'Erreur interne du bridge — vérifie audit.log'        if status >= 500

      raw.to_s.empty? ? "Erreur HTTP #{status}" : raw.to_s
    end

    def cmp_ver(a, b)
      pa = (a || '0').split('.').map(&:to_i)
      pb = (b || '0').split('.').map(&:to_i)
      [pa.size, pb.size].max.times do |i|
        va = pa[i] || 0
        vb = pb[i] || 0
        return -1 if va < vb
        return 1  if va > vb
      end
      0
    end
  end
end
